src/query/service/src/pipelines/processors/transforms/hash_join/spill_common.rs

```diff
@@ -23,17 +24,38 @@ use databend_common_expression::Expr;
 use databend_common_expression::FunctionContext;
 use databend_common_expression::HashMethodKind;
 use databend_common_functions::BUILTIN_FUNCTIONS;
+use databend_common_sql::plans::JoinType;
 
+use crate::pipelines::processors::transforms::hash_join::common::wrap_true_validity;
 use crate::pipelines::processors::transforms::hash_join::util::hash_by_method;
 
 pub fn get_hashes(
     func_ctx: &FunctionContext,
     block: &DataBlock,
     keys: &[Expr],
     method: &HashMethodKind,
+    join_type: Option<&JoinType>,
     hashes: &mut Vec<u64>,
 ) -> Result<()> {
-    let evaluator = Evaluator::new(block, func_ctx, &BUILTIN_FUNCTIONS);
+    let mut block = block.clone();
+    let mut evaluator = Evaluator::new(&block, func_ctx, &BUILTIN_FUNCTIONS);
+    if let Some(join_type) = join_type {
+        if matches!(
+            join_type,
+            JoinType::Left | JoinType::LeftSingle | JoinType::Full
+        ) {
+            let validity = Bitmap::new_constant(true, block.num_rows());
+            block = DataBlock::new(
+                block
+                    .columns()
+                    .iter()
+                    .map(|c| wrap_true_validity(c, block.num_rows(), &validity))
```

Comment:
If we don't add Bitmap, will the results be wrong ?

Reply:
Make `block`'s data_type align to key expr's data_type. Left-related join's build keys data_types are nullable.